File: RayTracer/Tracer/Scene.cs
using RayTracer.Common;
using RayTracer.Shape;
using RayTracer.Lighting;
using RayTracer.Material;
using RayTracer.BVH;
using RayTracer.Transformation;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;


namespace RayTracer.Tracer
{
    [Serializable]
    public class Scene
    {
        private Geometry[] geometries;
        private LinkedList<Transform> transforms = new LinkedList<Transform>();
        private Light[] lights;
        private List<Point3> vertices = new List<Point3>();
        private MyColor ambient;
        private Mat material;
        public Container Bvh;
        public MyColor defColor = new MyColor();

        public Scene()
        {
            Size = new Size();
            maxDepth = 5;
            transforms.AddFirst(new Scaling(new Point3(1, 1, 1)));
            material = new Mat();
            ambient = new MyColor(.2f, .2f, .2f);
            Attenuation = new Attenuation();
            OutputFilename = "default.bmp";
        }

        public String SceneFile { get; set; }
        public Scene(String scenefile)
            : this()
        {
            ParseCommand(scenefile);
        }

        public void ParseCommand(String scenefile)
        {
            SceneFile = scenefile;
            StreamReader filereader = new StreamReader(scenefile);
            string command;
            while ((command = filereader.ReadLine()) != null)
                ExecuteCommand(command);

            ConvertToArray();


            filereader.Close();
        }

        private void ConvertToArray()
        {
            lights = tempLights.ToArray();
            geometries = tempGeos.ToArray();
            tempGeos.Clear();
            tempLights.Clear();
            transforms.Clear();
            vertices.Clear();
            transforms.Clear();

        }

        string CleanCommand(string command)
        {
            command = command.Trim();
            command = Regex.Replace(command, @"\s+", " ");
            return command;
        }

        List<Light> tempLights = new List<Light>();
        List<Geometry> tempGeos = new List<Geometry>();
        public void ExecuteCommand(string fullcommand)
        {

            if (fullcommand.Contains('#'))
                return;

            fullcommand = CleanCommand(fullcommand);
            String[] words = fullcommand.Split(' ');
            String command = words[0];

            if (command.Equals("output"))
            {
                OutputFilename = words[1] + ".bmp";
                return;
            }

            float[] param = new float[words.Length - 1];
            for (int i = 0; i < param.Length; i++)
            {
                param[i] = float.Parse(words[i + 1]);
            }
            switch (command)
            {

                case "defColor":
                    defColor = new MyColor(param[0],param[1],param[2]);
                    break;
                case "size":
                    Size = new Size((int)param[0], (int)param[1]);
                    break;
                case "camera":
                    if (Camera.Instance != null) { break; }
                    Camera.Instance = new Camera(param);
                    ViewPlane.Instance = new ViewPlane(Size.Width, Size.Height);
                    break;
                case "maxdepth":
                    maxDepth = (int)param[0];
                    break;


                case "maxverts":
                    vertices = new List<Point3>((int)param[0]);
                    break;
                case "vertex":
                    vertices.Add(new Point3(param));
                    break;

                //geometry
                case "tri":
                    Geometry tri = CreateShape(Geometry.TYPE.TRIANGLE, param);
                    tempGeos.Add(tri);
                    break;

                case "sphere":
                    Geometry sphere = CreateShape(Geometry.TYPE.SPHERE, param);
                    tempGeos.Add(sphere);
                    break;

                //transforms 
                case "pushTransform":
                    transforms.AddFirst(Utils.DeepClone(transforms.First()));
                    break;
                case "popTransform":
                    transforms.RemoveFirst();
                    break;
                case "translate":
                    transforms.First().Matrix = Matrix.Mul44x44(transforms.First().Matrix, (new Translation(param)).Matrix);
                    break;
                case "scale":
                    transforms.First().Matrix = Matrix.Mul44x44(transforms.First().Matrix, (new Scaling(param)).Matrix);
                    break;
                case "rotate":
                    transforms.First().Matrix = Matrix.Mul44x44(transforms.First().Matrix, (new Rotation(param)).Matrix);
                    break;

                //material
                case "diffuse":
                    material.Diffuse = new MyColor(param[0], param[1], param[2]);
                    break;
                case "specular":
                    material.Specular = new MyColor(param[0], param[1], param[2]);
                    break;
                case "emission":
                    material.Emission = new MyColor(param[0], param[1], param[2]);
                    break;
                case "shininess":
                    material.Shininess = param[0];
                    break;
                case "refIndex":
                    material.RefractIndex = param[0];
                    break;
                case "refValue":
                    material.RefractValue = param[0];
                    break;


                //light
                case "attenuation":
                    Attenuation = new Attenuation(param);
                    break;
                case "ambient":
                    ambient = new MyColor(param);
                    break;
                case "directional":
                    tempLights.Add(new DirectionalLight(param));
                    break;
                case "point":
                    tempLights.Add(new PointLight(param));
                    break;

                default:
                    break;
            }
        }


        Geometry CreateShape(Geometry.TYPE type, float[] param)
        {
            Geometry geo;
            if (type == Geometry.TYPE.SPHERE)
            {
                geo = CreateSphere(param);
            }
            else //if (type == Geometry.TYPE.TRIANGLE)
            {
                geo = CreateTriangle(param);
            }
            ApplyTransform(geo);
            ApplyMaterial(geo);
            ApplyAmbient(geo);

            return geo;
        }

        Sphere CreateSphere(float[] param)
        {
            Sphere sphere = new Sphere(param);

            return sphere;
        }

        Triangle CreateTriangle(float[] param)
        {
            Point3 a = vertices[(int)param[0]];
            Point3 b = vertices[(int)param[1]];
            Point3 c = vertices[(int)param[2]];

            Triangle tri = new Triangle(a, b, c);

            return tri;
        }

        private void ApplyTransform(Geometry shape)
        {
            shape.Trans = Utils.DeepClone(transforms.First());
        }

        private void ApplyMaterial(Geometry shape)
        {
            shape.Material = Utils.DeepClone(material);
        }

        private void ApplyAmbient(Geometry shape)
        {
            shape.Ambient = Utils.DeepClone(ambient);
        }


        public Geometry[] Geometries
        {
            get { return geometries; }
            set { geometries = value; }
        }

        public Light[] Lights
        {
            get { return lights; }
        }

        int maxDepth;
        public int MaxDepth
        { get { return maxDepth; } }

        public string OutputFilename
        { get; set; }


        public Attenuation Attenuation
        { get; set; }

        public Size Size
        { get; set; }

        public void ShowInformation()
        {
            Console.WriteLine("=======Scene Information===========================");
            //Camera.ShowInformation();
            Console.WriteLine("Total Objects : " + Geometries.Length);
            Console.WriteLine("Total Lights  : " + Lights.Length);
            Console.WriteLine("Max Bounce    : " + MaxDepth);
            Console.WriteLine("Image Size    : " + Size.Width + " x " + Size.Height);
            Console.WriteLine("===================================================");
        }
    }
}

File: RayTracer - BVH - CUDA/RayTracerCUDA/Scene.cpp
#include "Scene.h"

Scene::Scene()
{
	att = Attenuation(new float[3]{ 1,0,0 });
}

Scene::Scene(std::string filename) :Scene()
{
	parseFile(filename);
}

void Scene::parseFile(std::string filename)
{
	std::string line;
	std::ifstream myfile(filename);
	if (myfile.is_open())
	{
		while (getline(myfile, line))
		{
			executeCommand(line);
		}
		myfile.close();
		vertices.clear();
	}
	else std::cout << "Unable to open file";
}

std::string Scene::CleanCommand(std::string command)
{
	command = std::regex_replace(command, std::regex("\t"), " "); //replace tab w/ space
	command = std::regex_replace(command, std::regex(" + "), " "); //replace multiple spaces w/ sigle space
	command = std::regex_replace(command, std::regex("^ +| +$"), ""); //replce leading and trailing space with nothing
	return command;
}

std::vector<std::string> Scene::splitString(std::string fullcommand, char delimiter) {
	std::vector<std::string> results;
	size_t pos = 0;
	std::string token;
	while ((pos = fullcommand.find(delimiter)) != std::string::npos) {
		token = fullcommand.substr(0, pos);
		results.push_back(token);
		fullcommand.erase(0, pos + 1);
	}
	results.push_back(fullcommand);
	fullcommand.erase(0, pos);
	return results;
}

void Scene::executeCommand(std::string fullcommand)
{
	if (fullcommand.compare("") == 0)
		return;
	if (fullcommand.find('#') != std::string::npos)
		return;

	fullcommand = CleanCommand(fullcommand);

	std::vector<std::string> words = splitString(fullcommand, ' ');
	std::string command = words[0];
	std::transform(command.begin(), command.end(), command.begin(), ::tolower);

	if (command.compare("output") == 0)
	{
		outFileName = words[1] + ".bmp";
		return;
	}

	std::vector<float> param;
	for (int i = 0; i < words.size() - 1; i++)
	{
		param.push_back(stof(words[i + 1]));
	}


	//	case "defColor":
	//		defColor = new MyColor(param[0], param[1], param[2]);
	//		break;
	if (command.compare("size") == 0) {
		size[0] = (int)param[0];
		size[1] = (int)param[1];
		return;
	}

	if (command.compare("camera") == 0) {

		Camera::Instance()->Init(&param[0]);
		ViewPlane::Instance()->Init(size[0], size[1]);
		return;
	}
	if (command.compare("maxdepth") == 0)
	{
		maxDepth = (int)param[0];
		return;
	}

	if (command.compare("maxvertex") == 0)
	{
		//no need		
		//vertices.reserve((int)param[0]);
		return;
	}
	if (command.compare("vertex") == 0)
	{
		vertices.push_back(std::make_shared<Vec3>(&param[0], 1));
		return;
	}

	//geometry
	if (command.compare("tri") == 0)
	{
		auto geo = createShape(&param[0]);
		geometries.push_back(geo);
		return;

	}


	//material

	if (command.compare("diffuse") == 0)
	{
		material.back()->diffuse = (MyColor(param[0], param[1], param[2]));
		return;
	}

	if (command.compare("specular") == 0)
	{
		material.back()->specular = (MyColor(param[0], param[1], param[2]));
		return;
	}
	if (command.compare("emission") == 0)
	{
		material.back()->emmission = (MyColor(param[0], param[1], param[2]));
		return;
	}
	if (command.compare("shininess") == 0)
	{
		material.back()->setShininess(param[0]);
		return;
	}

	if (command.compare("refindex") == 0)
	{
		material.back()->setrefIndex(param[0]);
		return;
	}
	if (command.compare("refvalue") == 0)
	{
		material.back()->setRefValue(param[0]);
		return;
	}

	//		//light
	if (command.compare("attenuation") == 0)
	{
		att = Attenuation(&param[0]);
		return;
	}
	if (command.compare("ambient") == 0)
	{
		material.back()->ambient = MyColor(param[0], param[1], param[2]);
		return;
	}
	if (command.compare("directional") == 0)
	{
		lights.push_back(std::make_shared<DirectionalLight>(new Vec3(param[0], param[1], param[2], 0), new MyColor(param[3], param[4], param[5])));
		return;
	}
	if (command.compare("point") == 0)
	{
		lights.push_back(std::make_shared<PointLight>(new Vec3(param[0], param[1], param[2], 1), new MyColor(param[3], param[4], param[5])));
		return;
	}
}

std::shared_ptr<Triangle> Scene::createShape(float * param)
{
	return std::make_shared<Triangle>(createTriangle(param));
}

Triangle Scene::createTriangle(float * param)
{
	Vec3 p[3];
	for (size_t i = 0; i < 3; i++)
	{
		p[i] = *vertices[(int)param[i]];
	}
	Triangle tri = Triangle(p[0], p[1], p[2]);
	applyMaterial(tri);
	return tri;
}


void Scene::applyMaterial(Triangle & shape)
{
	shape.mat = *material.back();
}

Scene::~Scene()
{
}
